rust/lance/src/index/vector/utils.rs

```diff
@@ -107,17 +107,72 @@ pub async fn maybe_sample_training_data(
     sample_size_hint: usize,
 ) -> Result<FixedSizeListArray> {
     let num_rows = dataset.count_rows(None).await?;
-    let batch = if num_rows > sample_size_hint {
+
+    let is_nullable = dataset
+        .schema()
+        .field(column)
+        .ok_or(Error::Index {
+            message: format!(
+                "Sample training data: column {} does not exist in schema",
+                column
+            ),
+            location: location!(),
+        })?
+        .nullable;
+
+    let batch = if num_rows > sample_size_hint && !is_nullable {
         let projection = dataset.schema().project(&[column])?;
         let batch = dataset.sample(sample_size_hint, &projection).await?;
         info!(
             "Sample training data: retrieved {} rows by sampling",
             batch.num_rows()
         );
         batch
+    } else if num_rows > sample_size_hint && is_nullable {
+        // Need to filter out null values
+        // Use a scan to collect row ids. Then sample from the row ids. Then do take.
+        let row_addrs = dataset
+            .scan()
+            .filter_expr(datafusion_expr::col(column).is_not_null())
+            .with_row_address()
+            .project::<&str>(&[])?
+            .try_into_batch()
+            .await?;
```

Comment:
@westonpace How expensive do you think this query is? This is filtering for non-null vectors and getting the row ids. Do you think there are easy optimizations we could do? If so, I'd like to capture that in a ticket.